### DISCUSSION TITLE
simulate keystrokes without the IsSimulated flag

### DISCUSSION THREAD

**Discussion by BlackCoyote** [TolikPylypchuk/SharpHook#122] [Q&A] [answered]
I was wondering if it is possible at all to send simulated keystrokes without having their `IsEventSimulated ` property set to true (as is reported by SharpHook's GlobalHook).

The reasoning is that some applications seem to straight up ignore any keystrokes that are simulated, making it completely impossible to create macros for them.

**Comment by TolikPylypchuk** [comment, marked as answer]
Hi! Thanks for the question! I don't think it's possible as the OS decides how the events are actually posted to applications. I also imagine that having control over this flag would give a little too much control to applications. I may be wrong though. I may look into it, but I don't think you'll have much luck with it.

I think it would actually be better (even if more difficult) to contact the maintainers of the applications that ignore simulated events and ask them to support these events.

**Reply by BlackCoyote** [reply]
I did come across [this StackOverflow post](https://stackoverflow.com/questions/54624221/simulate-physical-keypress-in-python-without-raising-lowlevelkeyhookinjected-0) that describes a similar situation. I assume that the `LowLevelKeyHookInjected` flag described there is the same that SharpHook calls `IsEventsimulated`? If so, maybe the response to that post could be relevant.

the question of whether or not it should be possible though is a tricky one indeed.
In my situation, I'm dealing with users of my application that uses SharpHook that are reporting that the keystrokes sent are being ignored by some big industry standard applications, like Pro Tools, which really don't have a reason to ignore the simulated keystrokes. Ideally they would indeed just change their application to stop ignoring simulated keys but if they actively made the choice to ignore them already, it's hard to imagine a few users would change their mind, so bypassing the simulated flag (if it were possible) would likely be the only way.

If actual harm could be done by simulating keystrokes without raising the simulated flag (and if it were possible at all), then I'd have to assume that darker and more obscure alternatives are already out there for bad actors, while allowing it to become more common place in public and open source projects would likely push applications that are somehow vulnerable to harm caused by secret simulated keystrokes to find more robust solutions rather than only letting the ability fall into the hands of more bad actors.

Ultimately though, If you'd rather not have your library provide the ability for that (if it were possible), then I'd fully understand and respect that. I just wanted to go a little more in-depth with my reasoning.

Thanks for the reply!

**Reply by TolikPylypchuk** [reply]
Yes, it appears that it really is possible to do, but it also appears that the process is much more involved. I've seen that people recommend installing a global hook which strips the "injected" flag from incoming events. This is obviously a problem in that _any_ simulated event will lose that flag, not just the ones raised by SharpHook.

And we're only talking about Windows here, don't forget about macOS and Linux as well, I have no idea whether it's possible on those OSes.

So yeah, sorry to disappoint you, but I won't add this feature or look into it further because it's technically complex and very niche. But the beauty of open source is that you can make your own version of SharpHook and use that if you wish, though that's going to be a lot of work as well :)

**Reply by BlackCoyote** [reply]
No worries! Thanks for taking the time to look into it and giving it some consideration.
